tests/DriverManagerTest.php

```diff
@@ -149,6 +149,28 @@ public function testDatabaseUrlPrimaryReplica(): void
         );
     }
 
+    public function testDatabaseUrlPrimaryReplicaWithServerVersionInUrl(): void
+    {
+        $serverVersion = 11.2;
+
+        $options = [
+            'driver' => 'pdo_mysql',
+            'primary' => ['url' => 'mysql://foo:bar@localhost:11211/baz?serverVersion=' . $serverVersion],
+            'replica' => [
+                'replica1' => ['url' => 'mysql://foo:bar@localhost:11211/baz_replica?serverVersion=' . $serverVersion],
+            ],
+            'wrapperClass' => PrimaryReadReplicaConnection::class,
+        ];
+
+        $conn = DriverManager::getConnection($options);
+
+        $params = $conn->getParams();
+        self::assertInstanceOf(PDO\MySQL\Driver::class, $conn->getDriver());
+
+        self::assertArrayHasKey('serverVersion', $params);
```

Comment:
What is the reason for this requirement? As I see it, without the fix, `$params` contains the 'primary' and 'replica.replica1' keys which contain the corresponding connection parameters including the server version:
```php
array(
  'driver' => 'pdo_mysql',
  'primary' => array(
    // ...
    'serverVersion' => '11.2',
  ),
  'replica' => array(
    'replica1' => array(
      // ...
      'serverVersion' => '11.2',
    ),
  ),
)
```

Why does it have to be also contained at the top level? What does top-level server version mean?

Reply:
https://github.com/doctrine/dbal/pull/4900#issuecomment-1015189401
Practically, without fix in this PR, Doctrine connects to DB to verify server version with every run of the code (while using replication).

Reply:
This is clear. The question is, why doesn't Doctrine use the server version specified in the corresponding connection parameters?

Reply:
The problem seems to be here: https://github.com/doctrine/dbal/blob/e839cecc812fd04ec6c39fc865302b6a322d967d/src/Connection.php#L398-L401

The connection doesn't expect the structure of its parameters changed by the extending class and expects the server version at the top level.

Since the PrimaryReplicaConnection changes the structure of parameters, it should change the way of detecting the platform accordingly. The closest relevant non-private method is:

https://github.com/doctrine/dbal/blob/e839cecc812fd04ec6c39fc865302b6a322d967d/src/Connection.php#L282

Reply:
I preferred changing the logic in DriverManager because it seemed to be the simplest way of fixing this issue.
Conenction is not Replication aware and it would require more changes to its logic, which I deemed undesirable.

Reply:
I’ll defer further review to @beberlei since he owns this part of the codebase.